src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,413 @@
+            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
+                throw new ArgumentOutOfRangeException(nameof(start));
+
+            Length = length;
+            _object = array;
+            _byteOffset = IntPtr.Zero.Add<T>(start);
+        }
+
+        /// <summary>
+        /// Creates a new span over the target unmanaged buffer.  Clearly this
+        /// is quite dangerous, because we are creating arbitrarily typed T's
+        /// out of a void*-typed block of memory.  And the length is not checked.
+        /// But if this creation is correct, then all subsequent uses are correct.
+        /// </summary>
+        /// <param name="ptr">An unmanaged pointer to memory.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when <typeparamref name="T"/> is reference type or contains pointers and hence cannot be stored in unmanaged memory.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        public unsafe Span(void* pointer, int length)
+        {
+            if (!SpanHelpers.IsReferenceFree<T>())
+                throw new ArgumentException(SR.Format(SR.Argument_InvalidTypeWithPointersNotSupported, typeof(T)));
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Length = length;
+            _object = null;
+            _byteOffset = new IntPtr(pointer);
+        }
+
+        /// <summary>
+        /// Create a new span over a portion of a regular managed object. This can be useful
+        /// if part of a managed object represents a "fixed array." This is dangerous because 
+        /// "length" is not checked, nor is the fact that "rawPointer" actually lies within the object.
+        /// </summary>
+        /// <param name="obj">The managed object that contains the data to span over.</param>
+        /// <param name="rawPointer">A reference to data within that object.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when the specified object is an array or string. Use the type-specific constructor
+        /// for these special cases.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when the specified object is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        public static Span<T> DangerousCreate(object obj, ref T rawPointer, int length)
```

Comment:
"ref T" is how it ended up in the final ApiReview: 

https://github.com/dotnet/apireviews/tree/master/2016/11-04-SpanOfT